scanner/enricher/nvd/nvd.go

```diff
@@ -0,0 +1,421 @@
+	resp, err := e.c.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("fetching NVD API results: %w", err)
+	}
+	zlog.Debug(ctx).
+		Int("status", resp.StatusCode).
+		Str("nvd_message", req.Header.Get("message")).
+		Msg("NVD response")
+	if resp.StatusCode != 200 {
+		_ = resp.Body.Close
+		return nil, fmt.Errorf("unexpected status code when querying %s: %d", req.URL.String(), resp.StatusCode)
+	}
+	return resp, nil
+}
+
+func parseResponse(resp *http.Response) (*schema.CVEAPIJSON20, error) {
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+	apiResp := new(schema.CVEAPIJSON20)
+	if err := json.NewDecoder(resp.Body).Decode(apiResp); err != nil {
+		return nil, fmt.Errorf("decoding API response: %w", err)
+	}
+	return apiResp, nil
+}
+
+// ParseEnrichment implements driver.EnrichmentUpdater.
+func (e *Enricher) ParseEnrichment(ctx context.Context, rc io.ReadCloser) ([]driver.EnrichmentRecord, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "enricher/nvd/Enricher/ParseEnrichment")
+	// Our Fetch method actually has all the smarts w/r/t to constructing the
+	// records, so this is just decoding in a loop.
+	defer func() {
+		_ = rc.Close()
+	}()
+	var err error
+	dec := json.NewDecoder(rc)
+	ret := make([]driver.EnrichmentRecord, 0, 1024) // Wild guess at initial capacity.
+	// This is going to allocate like mad, hold onto your butts.
+	for err == nil {
+		ret = append(ret, driver.EnrichmentRecord{})
+		err = dec.Decode(&ret[len(ret)-1])
+	}
+	zlog.Debug(ctx).
+		Int("count", len(ret)).
+		Msg("decoded enrichments")
+	if !errors.Is(err, io.EOF) {
+		return nil, err
+	}
+	return ret, nil
+}
+
+// This is a slightly more relaxed version of the validation pattern in the NVD
+// JSON schema: https://csrc.nist.gov/schema/nvd/api/2.0/source_api_json_2.0.schema
+//
+// It allows for "CVE" to be case-insensitive and for dashes and underscores
+// between the different segments.
+var cveRegexp = regexp.MustCompile(`(?i:cve)[-_][0-9]{4}[-_][0-9]{4,}`)
+
+// Enrich implements driver.Enricher.
+func (e *Enricher) Enrich(ctx context.Context, g driver.EnrichmentGetter, r *claircore.VulnerabilityReport) (string, []json.RawMessage, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "enricher/nvd/Enricher/Enrich")
+
+	// We return any CVSS blobs for CVEs mentioned in the free-form parts of the
+	// vulnerability.
+	m := make(map[string][]json.RawMessage)
+
+	erCache := make(map[string][]driver.EnrichmentRecord)
+	for id, v := range r.Vulnerabilities {
+		t := make(map[string]struct{})
+		ctx := zlog.ContextWithValues(ctx,
+			"vuln", v.Name)
+		for _, elem := range []string{
+			v.Description,
+			v.Name,
+			v.Links,
+		} {
+			for _, m := range cveRegexp.FindAllString(elem, -1) {
+				t[m] = struct{}{}
+			}
+		}
+		if len(t) == 0 {
+			continue
+		}
+		ts := make([]string, 0, len(t))
+		for m := range t {
+			ts = append(ts, m)
+		}
+		zlog.Debug(ctx).
+			Strs("cve", ts).
+			Msg("found CVEs")
+
+		sort.Strings(ts)
```

Comment:
nit: I think we should prefer https://pkg.go.dev/golang.org/x/exp/slices#Sort. Yes, it's in an experimental library, but once we adopt go1.21+, there will be a `slices` package with this same implementation of `Sort`, which is preferred over `sort.Strings` (see https://pkg.go.dev/sort#Strings)

Reply:
Let's just wait until we bump to 1.21 then.